File: modules/mauth-authenticator/src/main/java/com/mdsol/mauth/RequestAuthenticator.java
package com.mdsol.mauth;

import com.mdsol.mauth.exception.MAuthValidationException;
import com.mdsol.mauth.util.CurrentEpochTimeProvider;
import com.mdsol.mauth.util.EpochTimeProvider;
import com.mdsol.mauth.util.MAuthSignatureHelper;
import com.mdsol.mauth.utils.ClientPublicKeyProvider;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.charset.StandardCharsets;
import java.security.PublicKey;
import java.security.Security;
import java.util.Arrays;

public class RequestAuthenticator implements Authenticator {

  private static final Logger logger = LoggerFactory.getLogger(RequestAuthenticator.class);

  private final ClientPublicKeyProvider clientPublicKeyProvider;
  private final long requestValidationTimeoutSeconds;
  private final EpochTimeProvider epochTimeProvider;
  private final boolean v2OnlyAuthenticate;

  static {
    Security.addProvider(new BouncyCastleProvider());
  }

  /**
   * Uses
   * 10L as default value for request validation timeout
   * {@link com.mdsol.mauth.utils.ClientPublicKeyProvider} as the EpochTimeProvider
   *
   * @param clientPublicKeyProvider PublicKey provider
   */
  public RequestAuthenticator(ClientPublicKeyProvider clientPublicKeyProvider) {
    this(clientPublicKeyProvider, 10L);
  }

  /**
   * Uses
   * 10L as default value for request validation timeout
   * {@link com.mdsol.mauth.utils.ClientPublicKeyProvider} as the EpochTimeProvider
   *
   * @param clientPublicKeyProvider PublicKey provider
   * @param v2OnlyAuthenticate the flag to identify authenticate with protocol V2 only or not,
   *                  if true, clients will authenticate with protocol V2,
   *                  if false, clients will authenticate with only the highest protocol version (V2 or V1)
   */
  public RequestAuthenticator(ClientPublicKeyProvider clientPublicKeyProvider, boolean v2OnlyAuthenticate) {
    this(clientPublicKeyProvider, 10L, v2OnlyAuthenticate);
  }

  /**
   * Uses {@link com.mdsol.mauth.utils.ClientPublicKeyProvider} as the EpochTimeProvider
   *
   * @param clientPublicKeyProvider  PublicKey provider
   * @param requestValidationTimeoutSeconds timeout
   */
  public RequestAuthenticator(ClientPublicKeyProvider clientPublicKeyProvider, long requestValidationTimeoutSeconds) {
    this(clientPublicKeyProvider, requestValidationTimeoutSeconds, new CurrentEpochTimeProvider());
  }

  /**
   * Uses {@link com.mdsol.mauth.utils.ClientPublicKeyProvider} as the EpochTimeProvider
   *
   * @param clientPublicKeyProvider  PublicKey provider
   * @param requestValidationTimeoutSeconds timeout
   * @param v2OnlyAuthenticate the flag to identify authenticate with protocol V2 only or not,
   *                  if true, clients will authenticate with protocol V2,
   *                  if false, clients will authenticate with only the highest protocol version (V2 or V1)
   */
  public RequestAuthenticator(ClientPublicKeyProvider clientPublicKeyProvider,
      long requestValidationTimeoutSeconds, boolean v2OnlyAuthenticate) {
    this(clientPublicKeyProvider, requestValidationTimeoutSeconds, new CurrentEpochTimeProvider(), v2OnlyAuthenticate);
  }

  public RequestAuthenticator(ClientPublicKeyProvider clientPublicKeyProvider,
      long requestValidationTimeoutSeconds, EpochTimeProvider epochTimeProvider) {
    this(clientPublicKeyProvider, requestValidationTimeoutSeconds, epochTimeProvider, false);
  }

  public RequestAuthenticator(ClientPublicKeyProvider clientPublicKeyProvider,
      long requestValidationTimeoutSeconds, EpochTimeProvider epochTimeProvider, boolean v2OnlyAuthenticate) {
    this.clientPublicKeyProvider = clientPublicKeyProvider;
    this.requestValidationTimeoutSeconds = requestValidationTimeoutSeconds;
    this.epochTimeProvider = epochTimeProvider;
    this.v2OnlyAuthenticate = v2OnlyAuthenticate;
  }

  @Override
  public boolean authenticate(MAuthRequest mAuthRequest) {

    if (!(validateTime(mAuthRequest.getRequestTime()))) {
      final String message = "MAuth request validation failed because of timeout " + requestValidationTimeoutSeconds + "s";
      logger.error(message);
      throw new MAuthValidationException(message);
    }

    if (v2OnlyAuthenticate && !mAuthRequest.getMauthVersion().equals(MAuthVersion.MWSV2)) {
      final String message  = "The service requires mAuth v2 authentication headers.";
      logger.error(message);
      throw new MAuthValidationException(message);
    }

    PublicKey clientPublicKey = clientPublicKeyProvider.getPublicKey(mAuthRequest.getAppUUID());
    if (mAuthRequest.getMauthVersion().equals(MAuthVersion.MWSV2)) {
      boolean v2IsValidated = validateSignatureV2(mAuthRequest, clientPublicKey);
      if (v2OnlyAuthenticate) {
        return v2IsValidated;
      }
      else if (v2IsValidated) {
        return v2IsValidated;
      }
      else {
        return fallbackValidateSignatureV1(mAuthRequest, clientPublicKey);
      }
    }
    else {
      return validateSignatureV1(mAuthRequest, clientPublicKey);
    }

  }

  // Check epoch time is not older than specified interval.
  private boolean validateTime(long requestTime) {
    long currentTime = epochTimeProvider.inSeconds();
    return (currentTime - requestTime) < requestValidationTimeoutSeconds;
  }

  // check signature for V1
  private boolean validateSignatureV1 (MAuthRequest mAuthRequest, PublicKey clientPublicKey ) {

    logAuthenticationRequest(mAuthRequest);

    // Decrypt the signature with public key from requesting application.
    byte[] decryptedSignature = MAuthSignatureHelper.decryptSignature(clientPublicKey, mAuthRequest.getRequestSignature());

    // Recreate the plain text signature, based on the incoming request parameters, and hash it.
    try {
      byte[] messageDigest_bytes = MAuthSignatureHelper.generateUnencryptedSignature(
          mAuthRequest.getAppUUID(), mAuthRequest.getHttpMethod(), mAuthRequest.getResourcePath(),
          mAuthRequest.getMessagePayload(),
          String.valueOf(mAuthRequest.getRequestTime())
      );
      messageDigest_bytes = MAuthSignatureHelper.getHexEncodedDigestedString(messageDigest_bytes).getBytes(StandardCharsets.UTF_8);

      // Compare the decrypted signature and the recreated signature hashes.
      // If both match, the request was signed by the requesting application and is valid.
      return Arrays.equals(messageDigest_bytes, decryptedSignature);
    } catch (Exception ex) {
      final String message = "MAuth request validation failed for V1";
      logger.error(message, ex);
      throw new MAuthValidationException(message, ex);
    }
  }

  // check signature for V2
  private boolean validateSignatureV2 (MAuthRequest mAuthRequest, PublicKey clientPublicKey ) {

    logAuthenticationRequest(mAuthRequest);

    // Recreate the plain text signature, based on the incoming request parameters, and hash it.
    String unencryptedRequestString =
        MAuthSignatureHelper.generateStringToSignV2(
            mAuthRequest.getAppUUID(), mAuthRequest.getHttpMethod(), mAuthRequest.getResourcePath(),
            mAuthRequest.getQueryParameters(),
            mAuthRequest.getMessagePayload(),
            String.valueOf(mAuthRequest.getRequestTime())
        );

    // Compare the decrypted signature and the recreated signature hashes.
    try {
      return MAuthSignatureHelper.verifyRSA(unencryptedRequestString, mAuthRequest.getRequestSignature(), clientPublicKey);
    } catch (Exception ex) {
      final String message = "MAuth request validation failed for V2";
      logger.error(message, ex);
      throw new MAuthValidationException(message, ex);
    }
  }

  private boolean fallbackValidateSignatureV1(MAuthRequest mAuthRequest, PublicKey clientPublicKey) {
    boolean isValidated = false;
    if (mAuthRequest.getXmwsSignature() != null && mAuthRequest.getXmwsTime() != null) {
      MAuthRequest mAuthRequestV1 = new MAuthRequest(
          mAuthRequest.getXmwsSignature(),
          mAuthRequest.getMessagePayload(),
          mAuthRequest.getHttpMethod(),
          mAuthRequest.getXmwsTime(),
          mAuthRequest.getResourcePath(),
          mAuthRequest.getQueryParameters()
      );
      isValidated = validateSignatureV1(mAuthRequestV1, clientPublicKey);
      if (isValidated) {
        logger.warn("Completed successful authentication attempt after fallback to V1");
      }
    }
    return isValidated;
  }

  private void logAuthenticationRequest(MAuthRequest mAuthRequest) {
    final String msgFormat = "Mauth-client attempting to authenticate request from app with mauth app uuid %s using version %s.";
    logger.info(String.format(msgFormat, mAuthRequest.getAppUUID(), mAuthRequest.getMauthVersion().getValue()));
  }

}

File: modules/mauth-authenticator-akka-http/src/main/scala/com/mdsol/mauth/akka/http/RequestAuthenticator.scala
package com.mdsol.mauth.akka.http

import com.mdsol.mauth.{MAuthRequest, MAuthVersion}
import com.mdsol.mauth.exception.MAuthValidationException
import com.mdsol.mauth.scaladsl.Authenticator
import com.mdsol.mauth.scaladsl.utils.ClientPublicKeyProvider
import com.mdsol.mauth.util.{EpochTimeProvider, MAuthSignatureHelper}
import org.slf4j.{Logger, LoggerFactory}

import java.nio.charset.StandardCharsets
import java.security.PublicKey
import java.util
import scala.concurrent.duration._
import scala.concurrent.{ExecutionContext, Future, Promise}

class RequestAuthenticator(
  val publicKeyProvider: ClientPublicKeyProvider[Future],
  override val epochTimeProvider: EpochTimeProvider,
  v2OnlyAuthenticate: Boolean
)(implicit val executionContext: ExecutionContext)
    extends Authenticator[Future] {

  val logger: Logger = LoggerFactory.getLogger(classOf[RequestAuthenticator])

  def this(publicKeyProvider: ClientPublicKeyProvider[Future], epochTimeProvider: EpochTimeProvider)(implicit executionContext: ExecutionContext) =
    this(publicKeyProvider, epochTimeProvider, false)(executionContext)

  /** check if mauth v2 only authenticate is enabled or not
    *
    * @return True or false identifying if v2 only authenticate is enabled or not.
    */
  override val isV2OnlyAuthenticate: Boolean = v2OnlyAuthenticate

  /** Performs the validation of an incoming HTTP request.
    *
    * The validation process consists of recreating the mAuth hashed signature from the request data
    * and comparing it to the decrypted hash signature from the mAuth header.
    *
    * @param mAuthRequest Data from the incoming HTTP request necessary to perform the validation.
    * @return True or false indicating if the request is valid or not with respect to mAuth.
    */
  override def authenticate(mAuthRequest: MAuthRequest)(implicit requestValidationTimeout: Duration): Future[Boolean] = {
    val promise = Promise[Boolean]()
    if (!validateTime(mAuthRequest.getRequestTime)(requestValidationTimeout)) {
      val message = s"MAuth request validation failed because of timeout $requestValidationTimeout"
      logger.error(message)
      promise.failure(new MAuthValidationException(message))
    } else if (!validateMauthVersion(mAuthRequest, v2OnlyAuthenticate)) {
      val message = "The service requires mAuth v2 authentication headers."
      logger.error(message)
      promise.failure(new MAuthValidationException(message))
    } else {
      promise.completeWith(
        getPublicKey(mAuthRequest)
      )
    }
    promise.future
  }

  private def getPublicKey(mAuthRequest: MAuthRequest): Future[Boolean] = {
    publicKeyProvider.getPublicKey(mAuthRequest.getAppUUID).map {
      case None =>
        logger.error("Public Key couldn't be retrieved")
        false
      case Some(clientPublicKey) =>
        // Decrypt the signature with public key from requesting application.
        mAuthRequest.getMauthVersion match {
          case MAuthVersion.MWS =>
            logger.warn("MAuth v1 client was used to authenticate this request which is deprecated")
            validateSignatureV1(mAuthRequest, clientPublicKey)
          case MAuthVersion.MWSV2 =>
            val v2IsValidated = validateSignatureV2(mAuthRequest, clientPublicKey)
            if (v2OnlyAuthenticate)
              v2IsValidated
            else if (v2IsValidated)
              v2IsValidated
            else
              fallbackValidateSignatureV1(mAuthRequest, clientPublicKey)
        }
    }
  }

  // Check epoch time is not older than specified interval.
  private def validateTime(requestTime: Long)(requestValidationTimeout: Duration): Boolean =
    (epochTimeProvider.inSeconds - requestTime) < requestValidationTimeout.toSeconds

  // Check V2 header if only V2 is required
  private def validateMauthVersion(mAuthRequest: MAuthRequest, v2OnlyAuthenticate: Boolean): Boolean =
    !v2OnlyAuthenticate || mAuthRequest.getMauthVersion == MAuthVersion.MWSV2

  // check signature for V1
  private def validateSignatureV1(mAuthRequest: MAuthRequest, clientPublicKey: PublicKey): Boolean = {
    logAuthenticationRequest(mAuthRequest)
    val decryptedSignature = MAuthSignatureHelper.decryptSignature(clientPublicKey, mAuthRequest.getRequestSignature)
    // Recreate the plain text signature, based on the incoming request parameters, and hash it.
    try {
      val messageDigest_bytes = MAuthSignatureHelper.generateDigestedMessageV1(mAuthRequest).getBytes(StandardCharsets.UTF_8)

      // Compare the decrypted signature and the recreated signature hashes.
      // If both match, the request was signed by the requesting application and is valid.
      util.Arrays.equals(messageDigest_bytes, decryptedSignature)
    } catch {
      case ex: Exception =>
        val message = "MAuth request validation failed for V1."
        logger.error(message, ex)
        throw new MAuthValidationException(message, ex)
    }
  }

  // check signature for V2
  private def validateSignatureV2(mAuthRequest: MAuthRequest, clientPublicKey: PublicKey): Boolean = {
    logAuthenticationRequest(mAuthRequest)
    // Recreate the plain text signature, based on the incoming request parameters, and hash it.
    val unencryptedRequestString = MAuthSignatureHelper.generateStringToSignV2(mAuthRequest)

    // Compare the decrypted signature and the recreated signature hashes.
    try MAuthSignatureHelper.verifyRSA(unencryptedRequestString, mAuthRequest.getRequestSignature, clientPublicKey)
    catch {
      case ex: Exception =>
        val message = "MAuth request validation failed for V2."
        logger.error(message, ex)
        throw new MAuthValidationException(message, ex)
    }

  }

  private def fallbackValidateSignatureV1(mAuthRequest: MAuthRequest, clientPublicKey: PublicKey): Boolean = {
    var isValidated = false
    if (mAuthRequest.getMessagePayload == null) {
      logger.warn("V1 authentication fallback is not available because the full request body is not available in memory.")
    } else if (mAuthRequest.getXmwsSignature != null && mAuthRequest.getXmwsTime != null) {
      val mAuthRequestV1 = new MAuthRequest(
        mAuthRequest.getXmwsSignature,
        mAuthRequest.getMessagePayload,
        mAuthRequest.getHttpMethod,
        mAuthRequest.getXmwsTime,
        mAuthRequest.getResourcePath,
        mAuthRequest.getQueryParameters
      )
      isValidated = validateSignatureV1(mAuthRequestV1, clientPublicKey)
      if (isValidated) {
        logger.warn("Completed successful authentication attempt after fallback to V1")
      }
    }
    isValidated
  }

  private def logAuthenticationRequest(mAuthRequest: MAuthRequest): Unit = {
    val msgFormat = "Mauth-client attempting to authenticate request from app with mauth app uuid %s using version %s."
    logger.info(String.format(msgFormat, mAuthRequest.getAppUUID, mAuthRequest.getMauthVersion.getValue))
  }
}
